### ISSUE TITLE
more advanced configurations options

### ISSUE THREAD

**Issue by jszigetvari** [OndrejHome/ansible.pcs-modules-2#7]
Dear Ondrej,

I'm doing a pet project the goal of which would be to create a more or less rudimentary Ansible setup script for a syslog-ng HA cluster with DRBD, on top of CentOS 7.
In search for Ansible pcs cluster modules, yours seemed to be the prime choice.

I'm following a cluster setup guide of a friend, and creating the tasks of the playbook along those steps.
I'm more like a beginner in pacemaker-based clustering, so I'd need some help.

There seem to be some steps that I can't currently automate with your pcs_modules.
Could you please verify this assumption? The commands are the ones below:

`$ pcs cluster start --all`
`$ pcs resource master ResourceAClone ResourceA master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true`
`$ pcs resource defaults resource-stickiness=100`
`$ pcs constraint colocation add ResourceB with ResourceAClone INFINITY with-rsc-role=Master`

If these commands are not doable with the current version of your modules, then how big of an effort would it take to implement them?

Thank you in advance for your help, and thanks for the work you do!

**Comment by OndrejHome** [comment]
Hi Janos ( @jszigetvari ),

For creating a basic pacemaker cluster I have a role named 'OndrejHome.ha-cluster-pacemaker' that installs the needed packages, setups cluster, starts it and can configure fencing device.
Instead of doing the `pcs cluster start --all` I use the following task in ansible that has same effect.
https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/tasks/main.yml#L57
~~~
- name: Start cluster services on all nodes
  service: name={{ item }} state=started
  with_items:
    - pacemaker 
    - corosync
~~~

As for creating a Master/Slave resource  you can use the `pcs_resource` module that can handle it. It only requires to provide a valid syntax that creates the Master/Slave resource. For example:
~~~
- pcs_resource: name="ResourceAClone" resource_type="ResourceAtype" options="someOption=value someOption2=value2 --master master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true"
~~~
(Example uses some options that belongs to `ResourceA` first and then after `--master` keyword it specifies all options for Master/Slave resource)

Colocation constraint can be created with `pcs_colocation_constraint`. Following your example it would like this:
~~~
- pcs_colocation_constraint: resource1='ResourceB' resource2='ResourceAClone' resource2_role='Master'
~~~

Current modules are not implementing set of global defaults for resources or resource operations, but you can workaround this by specifying the needed options directly in the `pcs_resource` options. So you can have values that you need, but this will need to be in current version added/specified manually  for each resource.

To summarise what can be improved:
- [ ] adding example of Master/Slave resource into `pcs_resource` examples
- [ ] ability to setup 'resource defaults` and 'resource operation defaults'

Adding example is not a problem, adding ability for resource default will take some time (will have look at this some upcoming weekends) as it will most probably need a new module and some testing. I will keep this issue open to track this and let you know.

Please let me know if the above examples helped you to achieve the configuration that you want.

**Comment by frank3427** [comment]
I am also interested in;
pcs resource defaults resource-stickiness=100

**Comment by jszigetvari** [comment]
Dear Ondrej,

Thank you for your answers.
I have made the necessary modifications to my playbook, the next step will be to test it.

I have one question though. If we look at one of the original commands I asked:

`$ pcs resource master ResourceAClone ResourceA master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true`

Your suggestion follows:

`- pcs_resource: name="ResourceAClone" resource_type="ResourceAtype" options="someOption=value someOption2=value2 --master master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true"`

Your suggestion however doesn't bear any hints that the clone is actually a clone for ResourceA. In case of the original version it's clearly visible.
Won't your format lead to any confusion within pcsd?
Or is there any way to explicitly state that? Is it needed at all?

Thank you for suggesting the workarounds. Hopefully the modified playbooks will work fine.

**Comment by OndrejHome** [comment]
Hi Janos,

Let me try to show this through examples - I will use the `ocf:pacemaker:Dummy` as example type of resource.

Your approach has basically 2 steps:
~~~
# pcs resource create ResourceA ocf:pacemaker:Dummy fake=some_value

 Resource: ResourceA (class=ocf provider=pacemaker type=Dummy)
  Attributes: fake=some_value

# pcs resource master ResourceAClone ResourceA master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true

 Master: ResourceAClone
  Meta Attrs: master-node-max=1 clone-max=2 notify=true master-max=1 clone-node-max=1
  Resource: ResourceA (class=ocf provider=pacemaker type=Dummy)
   Attributes: fake=some_value
~~~

With `pcs_resource` module here I'm proposing a single step that will get us to "same" result (except of the name of the final Master/Slave resource which will have suffix `-master`).
~~~
# pcs resource create ResourceAClone ocf:pacemaker:Dummy fake=some_value --master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true

 Master: ResourceAClone-master
  Meta Attrs: clone-max=2 clone-node-max=1 master-max=1 master-node-max=1 notify=true
  Resource: ResourceAClone (class=ocf provider=pacemaker type=Dummy)
   Attributes: fake=some_value
~~~
This second approach with only one command is something that can be done with `pcs_resource` modules using definition below.
~~~
- pcs_resource: name="ResourceAClone" resource_type="ocf:pacemaker:Dummy" options="fake=some_value master master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true"
~~~

Please let me know if the above makes sense now.

PS: in some newer versions of pcs the `--master` option may throw warning, in such case use the `master` instead as suggested in warning. I will check if this can be also somehow fixed in the `pcs_resource` code as that warning is harmless.

**Comment by OndrejHome** [comment]
Update regarding the use of `--master`: the warning is ignored correctly, but there is slight difference in syntax, so based on your version of pcs one of both of below should work same way:
~~~
- pcs_resource: name="ResourceAClone" resource_type="ocf:pacemaker:Dummy" options="fake=some_value --master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true"

- pcs_resource: name="ResourceAClone" resource_type="ocf:pacemaker:Dummy" options="fake=some_value master master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true"
~~~

**Comment by jszigetvari** [comment]
Dear Ondrej,

Thank you so far for trying to help me.
My problem at the moment seems to be that ResourceA is really a DRBD device, and when creating it, op is set to monitor, like this:

`pcs -f drbd_cfg resource create DRBDSharedStorage ocf:linbit:drbd drbd_resource=shareddata op monitor interval=60s`

`pcs -f drbd_cfg resource master DRBDSharedStorageClone DRBDSharedStorage master-
max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true`

This makes the creation of the -master resource in one step difficult, as AFAIK it's EITHER op monitor OR master, but not both. Or am I wrong with that assumption?
Or should I simply skip "op monitor interval=60s"?

Furthermore, when setting up the constraints I use this yaml config:

```    - name: 
      pcs_constraint_colocation: >
        state="present"
        resource1="SyslogFS"
        resource2="DRBDSharedStorage-master"
        resource2_role="Master"
        score="INFINITY"

    - name: 
      pcs_constraint_order: >
        state="present"
        resource1="DRBDSharedStorage-master"
        resource1_action="promote"
        resource2="SyslogFS"```

And in turn, I get the following nice error message:

```TASK [pcs_constraint_order] *****************************************************************************************************************************************************************************************************************
changed: [node2]
fatal: [node1]: FAILED! => {"changed": false, "error": "Error: duplicate constraint already exists, use --force to override\n  promote DRBDSharedStorage-master then start SyslogFS (kind:Mandatory) (id:order-StorageSyslog-master-SyslogFS-Mandatory)\n", "failed": true, "msg": "Failed to create constraint with cmd: 'pcs constraint order promote DRBDSharedStorage-master then start SyslogFS kind=Mandatory symmetrical=true'", "output": ""}```

One other thing: given that `pcs resource defaults resource-stickiness=100` is not available, I tried adding resource-stickiness=100 to every pcs_resource-based resource creation.
Unfortunately it broke resource creations with `ocf:heartbeat:IPaddr2`, `ocf:heartbeat:syslog-ng` and `ocf:heartbeat:Filesystem`
If most of the resource types don't honor it, then where should I use it?

Also, the original guide builds upon the tactics of creating a separate cluster config file, carrying out modifications on that separate file, then committing the changes as a single transaction.
Currectly this course of action is not possible with your modules. :(

**Comment by OndrejHome** [comment]
Hi Janos,

You can create Master/Slave resource with changes to various operations in single command, please see the example below on the Dyummy resource.

~~~
# pcs resource create ResourceAClone ocf:pacemaker:Dummy fake=some_value --master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true op monitor interval=60s

 Master: ResourceAClone-master
  Meta Attrs: clone-max=2 clone-node-max=1 master-max=1 master-node-max=1 notify=true
  Resource: ResourceAClone (class=ocf provider=pacemaker type=Dummy)
   Attributes: fake=some_value
   Operations: monitor interval=60s (ResourceAClone-monitor-interval-60s)
               start interval=0s timeout=20 (ResourceAClone-start-interval-0s)
               stop interval=0s timeout=20 (ResourceAClone-stop-interval-0s)
~~~

Additionally you can also add the 'resource-stickiness' also in single command using one of the commands below.
~~~
# pcs resource create ResourceAClone ocf:pacemaker:Dummy fake=some_value meta resource-stickiness=99 --master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true op monitor interval=60s
# pcs resource create ResourceAClone ocf:pacemaker:Dummy fake=some_value --master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true op monitor interval=60s meta resource-stickiness=99

 Master: ResourceAClone-master
  Meta Attrs: clone-max=2 clone-node-max=1 master-max=1 master-node-max=1 notify=true resource-stickiness=99
  Resource: ResourceAClone (class=ocf provider=pacemaker type=Dummy)
   Attributes: fake=some_value
   Operations: monitor interval=60s (ResourceAClone-monitor-interval-60s)
               start interval=0s timeout=20 (ResourceAClone-start-interval-0s)
               stop interval=0s timeout=20 (ResourceAClone-stop-interval-0s)
~~~

I'm curios what went wrong with `ocf:heartbeat:IPaddr2`, `ocf:heartbeat:syslog-ng` and `ocf:heartbeat:Filesystem`. Please note that you should put the `resource-stickiness=XXX` after the `meta` keyword as shown below.
~~~
# pcs resource create testIP IPaddr2 ip=192.168.22.33 meta resource-stickiness=100

 Resource: testIP (class=ocf provider=heartbeat type=IPaddr2)
  Attributes: ip=192.168.22.33
  Meta Attrs: resource-stickiness=100
  Operations: monitor interval=10s timeout=20s (testIP-monitor-interval-10s)
              start interval=0s timeout=20s (testIP-start-interval-0s)
              stop interval=0s timeout=20s (testIP-stop-interval-0s)

~~~

As for the constraints I can see one thing that may cause issues here. Seems that you are calling the modules for each node (node1 and node2). As the cluster configuration is for whole cluster it is enough to run it on one of the nodes, for example by using `run_once: true` directive. From the output you have provided it looks that both nodes attempted to configure the constraint at same time. In such case I would expect that both nodes detected that constraint is not present and then both of them tried to create it. Logically only one would succeed and other would complain that constraint already exists. Could you try to run this only from one of the nodes?

> Also, the original guide builds upon the tactics of creating a separate cluster config file, carrying out modifications on that separate file, then committing the changes as a single transaction.

I'm not sure what is the benefit of committing all changes in one transaction compared to making them gradually. In both cases the result should be same. Is there something that cannot be done gradually and would require a transaction to be used? (I would expect that some combination of constraints might be better to be applied through single transaction rather than gradually, but I haven't seen yet a strong need for this, if you have some concrete use case that works with transaction but not gradually that would be useful here)

Also if the guide that your are following is publicly available could you share a link to it here so I can check it?

In the meantime it seems that the 'resource defaults' and 'resource operation defaults' should be easy to implement into new module(s). I will let you know once some testing module(s) are available.

**Comment by OndrejHome** [comment]
@jszigetvari, @frank3427,
In the 'develop' branch I have pushed the new module called `pcs_resource_defaults` that can manipulate resource defaults and resource operation defaults. Please let me know if you have any feedback to this new module. If there are no big issues it will be included in next version of pcs-modules-2.

**Comment by jszigetvari** [comment]
@OndrejHome 
Thank you!
I downloaded the new branch, and the `pcs_resource_defaults` module did seem to work okay for me, although I did not do any far-fetched testing. I just verified that the default stickiness was set up successfully by the module.

**Comment by OndrejHome** [comment]
Hi Janos (@jszigetvari )
I have merged the changes into 'master' branch so you can now use the new version of 'OndrejHome.pcs-modules-2' ('v12') directly from ansible galaxy.

I will keep this Issue open to wait for your feedback. Please let me know if you encounter any other issues or if the issue can be closed.

**Comment by jszigetvari** [comment]
Hi Ondrej,

all-in-all things look good for me.
At the moment, I don't have any further questions.
Thank you for your help!
